tests/templates/test_subroutines.py

```diff
@@ -1207,3 +1208,284 @@ def circuit():
             return [qml.expval(qml.PauliZ(wires=i)) for i in range(n_wires)]
 
         assert np.allclose(circuit(), expectation)
+
+
+class TestPermute:
+    """Tests for the Permute template from the pennylane.templates.subroutine module."""
+
+    def test_invalid_inputs_qnodes(self):
+        """Tests if errors are thrown for invalid permutations with QNodes."""
+
+        dev = qml.device("default.qubit", wires=4)
+
+        with pytest.raises(ValueError):
+
+            @qml.qnode(dev)
+            def permute_qubits():
+                qml.templates.Permute([0], wires=[0])
+                return qml.expval(qml.PauliZ(0))
+
+            permute_qubits()
+
+        with pytest.raises(ValueError):
+
+            @qml.qnode(dev)
+            def permute_qubits():
+                qml.templates.Permute([0, 1, 2], wires=[0, 1, 2, 3])
+                return qml.expval(qml.PauliZ(0))
+
+            permute_qubits()
```

Comment:
Each of these test cases should likely be a separate test function, so that all are tested even if the first fails.

Reply:
Is there a way of testing for error handling using the pytest parameterization and multiple input/output pairs (as is done below for the other tests)?

Reply:
Actually it was more straightforward than I expected to do this - it looks much tidier now.